src/librustc_borrowck/borrowck/mir/mod.rs

```diff
@@ -89,3 +174,127 @@ impl<'b, 'a: 'b, 'tcx: 'a> MirBorrowckCtxt<'b, 'a, 'tcx> {
+                }
+            }
+            _ => true
+        }
+    }
+
+    fn on_all_children_bits<'a, 'tcx, F>(
+        tcx: TyCtxt<'a, 'tcx, 'tcx>,
+        mir: &Mir<'tcx>,
+        move_data: &MoveData<'tcx>,
+        move_path_index: MovePathIndex,
+        each_child: &mut F)
+        where F: FnMut(MovePathIndex)
+    {
+        each_child(move_path_index);
+
+        if is_terminal_path(tcx, mir, move_data, move_path_index) {
+            return
+        }
+
+        let mut next_child_index = move_data.move_paths[move_path_index].first_child;
+        while let Some(child_index) = next_child_index {
+            on_all_children_bits(tcx, mir, move_data, child_index, each_child);
+            next_child_index = move_data.move_paths[child_index].next_sibling;
+        }
+    }
+    on_all_children_bits(tcx, mir, move_data, move_path_index, &mut each_child);
+}
+
+fn drop_flag_effects_for_function_entry<'a, 'tcx, F>(
+    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    mir: &Mir<'tcx>,
+    move_data: &MoveData<'tcx>,
+    mut callback: F)
+    where F: FnMut(MovePathIndex, DropFlagState)
+{
+    for i in 0..(mir.arg_decls.len() as u32) {
+        let lvalue = repr::Lvalue::Arg(i);
+        let move_path_index = move_data.rev_lookup.find(&lvalue);
+        on_all_children_bits(tcx, mir, move_data,
+                             move_path_index,
+                             |moi| callback(moi, DropFlagState::Live));
+    }
+}
+
+fn drop_flag_effects_for_location<'a, 'tcx, F>(
+    tcx: TyCtxt<'a, 'tcx, 'tcx>,
+    mir: &Mir<'tcx>,
+    move_data: &MoveData<'tcx>,
+    loc: Location,
+    mut callback: F)
+    where F: FnMut(MovePathIndex, DropFlagState)
+{
+    debug!("drop_flag_effects_for_location({:?})", loc);
+
+    // first, move out of the RHS
+    for mi in &move_data.loc_map[loc] {
+        let path = mi.move_path_index(move_data);
+        debug!("moving out of path {:?}", move_data.move_paths[path]);
+
+        // don't move out of non-Copy things
+        if let MovePathContent::Lvalue(ref lvalue) = move_data.move_paths[path].content {
+            let ty = mir.lvalue_ty(tcx, lvalue).to_ty(tcx);
+            let empty_param_env = tcx.empty_parameter_environment();
+            if !ty.moves_by_default(tcx, &empty_param_env, DUMMY_SP) {
+                continue;
+            }
+        }
+
+        on_all_children_bits(tcx, mir, move_data,
+                             path,
+                             |moi| callback(moi, DropFlagState::Dead))
+    }
+
+    let bb = mir.basic_block_data(loc.block);
+    match bb.statements.get(loc.index) {
+        Some(stmt) => match stmt.kind {
+            repr::StatementKind::Assign(ref lvalue, _) => {
+                debug!("drop_flag_effects: assignment {:?}", stmt);
+                on_all_children_bits(tcx, mir, move_data,
+                                     move_data.rev_lookup.find(lvalue),
+                                     |moi| callback(moi, DropFlagState::Live))
```

Comment:
Nit: I think I would prefer the name `Initialized` or `Moved` to `Live` and `Dead`. `Live` in particular suggests liveness to me, and merely being initialized isn't enough for that :)